src/GraphQL/Server.hs

```diff
@@ -77,11 +76,21 @@ infixr 8 :<>
 data a :<|> b = a :<|> b
 infixr 8 :<|>
 
+
+-- Result collects errors and values at the same time unless a handler
+-- tells us to bail out in which case we stop the processing
+-- immediately.
+-- TODO: Make QueryError a Monoid and aggregate according to that.
+data Result = Result [QueryError] GValue.Value
+ok :: GValue.Value -> Result
+ok v = Result [] v
+
+
 -- TODO instead of SelectionSet we want something like
 -- NormalizedSelectionSet which has query fragments etc. resolved.
-class (MonadThrow m, MonadIO m) => HasGraph m a where
+class (Monad m) => HasGraph m a where
```

Comment:
You don't need to say `Monad m` here, and I don't think you should. In general, it's a good idea to put constraints where they are actually needed.

The only reason you might want to put it here is for `HasGraph m a` instances to always imply `Monad m` (`Eq a => Ord a` being the canonical example). But we already explicitly say `Monad m` on each instance.